src/string_decoder.cc

```diff
@@ -0,0 +1,334 @@
+              // We either have more then 4 trailing bytes (which means
+              // the current character would not be inside the range for
+              // valid Unicode, and in particular cannot be represented
+              // through JavaScript's UTF-16-based approach to strings), or the
+              // current buffer does not contain the start of an UTF-8 character
+              // at all. Either way, this is invalid UTF8 and we can just
+              // let the engine's decoder handle it.
+              state_[kBufferedBytes] = 0;
+              break;
+            }
+          } else {
+            // Found the first byte of a UTF-8 character. By looking at the
+            // upper bits we can tell how long the character *should* be.
+            if ((data[i] & 0xE0) == 0xC0) {
+              state_[kMissingBytes] = 2;
+            } else if ((data[i] & 0xF0) == 0xE0) {
+              state_[kMissingBytes] = 3;
+            } else if ((data[i] & 0xF8) == 0xF0) {
+              state_[kMissingBytes] = 4;
+            } else {
+              // This lead byte would indicate a character outside of the
+              // representable range.
+              state_[kBufferedBytes] = 0;
+              break;
+            }
+
+            if (BufferedBytes() >= MissingBytes()) {
+              // Received more or exactly as many trailing bytes than the lead
+              // character would indicate. In the "==" case, we have valid
+              // data and don't need to slice anything off;
+              // in the ">" case, this is invalid UTF-8 anyway.
+              state_[kMissingBytes] = 0;
+              state_[kBufferedBytes] = 0;
+            }
+
+            state_[kMissingBytes] -= state_[kBufferedBytes];
+            break;
+          }
+        }
+      } else if (Encoding() == UCS2) {
+        if ((nread % 2) == 1) {
+          // We got half a codepoint, and need the second byte of it.
+          state_[kBufferedBytes] = 1;
+          state_[kMissingBytes] = 1;
+        } else if ((data[nread - 1] & 0xFC) == 0xD8) {
+          // Half a split UTF-16 character.
+          state_[kBufferedBytes] = 2;
+          state_[kMissingBytes] = 2;
+        }
+      } else if (Encoding() == BASE64) {
+        state_[kBufferedBytes] = nread % 3;
+        if (state_[kBufferedBytes] > 0)
+          state_[kMissingBytes] = 3 - BufferedBytes();
+      }
+
+      if (BufferedBytes() > 0) {
+        // Copy the requested number of buffered bytes from the end of the
+        // input into the incomplete character buffer.
+        nread -= BufferedBytes();
+        *nread_ptr -= BufferedBytes();
+        memcpy(IncompleteCharacterBuffer(), data + nread, BufferedBytes());
+      }
+
+      if (nread > 0) {
+        if (!MakeString(isolate, data, nread, Encoding()).ToLocal(&body))
+          return MaybeLocal<String>();
+      } else {
+        body = String::Empty(isolate);
+      }
+    }
+
+    if (!prepend.IsEmpty()) {
```

Comment:
Arguably reads a bit more natural  if you invert (or un-invert really) the logic:
```c++
if (prepend.IsEmpty()) return body;
return String::Concat(prepend, body);
```